### PR TITLE
For deployment 2024 06 07

### PR DESCRIPTION
**Pull request by xueh2**
update the environment.yml
fix cmake files to find more packages
install .cmake files

### CODE REVIEW

File: environment.yml
Comment (hansenms) [inline environment.yml:1]: Please keep the comments properly aligned in this file.

File: environment.yml [hunk 1]
```diff
@@ -26,53 +26,52 @@ dependencies:
...
+  - eigen=3.4.0
   - fftw=3.3.9
-  - gadgetron::gadgetron-python=1.4.1
-  - gcc_linux-64=9.4.0                  # dev
-  - gdb=11.1                            # dev
-  - gmock=1.10.0                        # dev
-  - gxx_linux-64=9.4.0                  # dev
-  - gtest=1.10.0                        # dev
-  - h5py=3.2.1
+  - gadgetron-python=1.4.1
```
Comment (hansenms) [inline environment.yml:33]: How come you have removed the channel specification on these packages?

File: CMakeLists.txt
Comment (hansenms) [inline CMakeLists.txt:1]: You should bump the patch version in this file.

File: apps/gadgetron/connection/Loader.h [hunk 2]
```diff
@@ -38,6 +38,7 @@ namespace Gadgetron::Server::Connection {
 
         template<class FACTORY>
         FACTORY& load_factory(const std::string &prefix, const std::string &classname, const std::string &dll) {
+            GINFO_STREAM("loading " << prefix << " - " << classname << " from the dll " << dll);
```
Comment (hansenms) [inline apps/gadgetron/connection/Loader.h:41]: Don't think you need this?

File: cmake/FindCUDA_advanced.cmake [hunk 3]
```diff
@@ -13,6 +13,8 @@ if(CUDA_FOUND)
   set(CUDA_NVCC_FLAGS75 "-gencode arch=compute_75,code=sm_75")
   set(CUDA_NVCC_FLAGS8 "-gencode arch=compute_80,code=sm_80")
   set(CUDA_NVCC_FLAGS86 "-gencode arch=compute_86,code=sm_86")
+  set(CUDA_NVCC_FLAGS87 "-gencode arch=compute_87,code=sm_87")
```
Comment (hansenms) [inline cmake/FindCUDA_advanced.cmake:16]: Should we discontinue some of the older ones here to keep the compile time down, do er really need to go back to 6?

File: conda/meta.yaml [hunk 4]
```diff
@@ -11,71 +11,71 @@ source:
...
     - gxx_linux-64=9.4.0                         # [linux64]
-    - gtest=1.10.0
-    - h5py=3.2.1
+    - gtest=1.14.0
+    - h5py=3.7.0
     - hdf5=1.10.6
     - howardhinnant_date=3.0.1
-    - ismrmrd::ismrmrd=1.13.7
-    - ismrmrd::ismrmrd-python>=1.9.8
+    - ismrmrd=1.13.7
```
Comment (hansenms) [inline conda/meta.yaml:31]: Channel specifiers are missing here too, might be fine by now, but there were a lot of versions floating around at some point.

File: gadgets/mri_core/generic_recon_gadgets/GenericImageReconGadget.cpp [hunk 5]
```diff
@@ -101,6 +101,10 @@ namespace Gadgetron {
     {
         try
         {
+            // every image requires these two fields
```
Comment (hansenms) [inline gadgets/mri_core/generic_recon_gadgets/GenericImageReconGadget.cpp:104]: Seems odd that every image coming out of the Gadgetron should need those fields?

File: gadgets/mri_noncartesian/GriddingReconGadgetBase.hpp [hunk 6]
```diff
@@ -95,8 +95,8 @@ template<template<class> class ARRAY> 	int GriddingReconGadgetBase<ARRAY>::proce
 			
 			if (buffer->headers_[0].trajectory_dimensions == 3){
 				auto traj_dcw = separate_traj_and_dcw(&trajectory);
-				dcw = boost::make_shared<ARRAY<float>>(std::get<1>(traj_dcw).get());
-				traj = boost::make_shared<ARRAY<floatd2>>(std::get<0>(traj_dcw).get());
+				dcw = boost::make_shared<ARRAY<float>>(*std::get<1>(traj_dcw).get());
```
Comment (hansenms) [inline gadgets/mri_noncartesian/GriddingReconGadgetBase.hpp:98]: This seems odd, why are you dereferencing here?

File: test/integration/run_gadgetron_test.py [hunk 7]
```diff
@@ -619,8 +619,9 @@ def stdout_compliance_action(cont, **state):
         files.append(os.path.join(args.test_folder, 'gadgetron.log.out'))
 
         for file in files:
-            if os.stat(file).st_size != 0:
-                raise RuntimeError(f"stdout is not empty as indicated by {file}")
+            if os.path.isfile(file):
```
Comment (hansenms) [inline test/integration/run_gadgetron_test.py:622]: How would it not be a file?

File: toolboxes/core/cpu/math/hoNDArray_linalg.cpp [hunk 8]
```diff
@@ -590,5 +590,51 @@ template void linFit(const hoNDArray<double>& x, const hoNDArray<double>& y, dou
 template void linFit(const hoNDArray< std::complex<float> >& x, const hoNDArray< std::complex<float> >& y, std::complex<float>& a, std::complex<float>& b);
 template void linFit(const hoNDArray< std::complex<double> >& x, const hoNDArray< std::complex<double> >& y, std::complex<double>& a, std::complex<double>& b);
 
+template <typename T>
+void polyFit2(const hoNDArray<T>& x, const hoNDArray<T>& y, T& a, T& b, T& c)
```
Comment (hansenms) [inline toolboxes/core/cpu/math/hoNDArray_linalg.cpp:594]: We should probably have some kind of unit test for this?

**Review by hansenms — CHANGES_REQUESTED** [review]
Looks good overall. Some minor suggestions.

File: environment.yml (hunk 1 above)
Comment (xueh2) [reply environment.yml:33]: they are not needed. there are no ambiguity for the ismrmrd and gadgetron packages.

File: CMakeLists.txt
Comment (xueh2) [reply CMakeLists.txt:1]: fixed. bump to 4.6.1

File: apps/gadgetron/connection/Loader.h (hunk 2 above)
Comment (xueh2) [reply apps/gadgetron/connection/Loader.h:41]: No harm to print out these information ? It gives hints if loading classes failed and does not occupy much lines in the log.

File: cmake/FindCUDA_advanced.cmake (hunk 3 above)
Comment (xueh2) [reply cmake/FindCUDA_advanced.cmake:16]: retire compute_6x and _7x.

File: conda/meta.yaml (hunk 4 above)
Comment (xueh2) [reply conda/meta.yaml:31]: It is ok now. For future, all ismrmrd should only be published as conda ? If so, it is safe for future versions, without the channel spec.

File: gadgets/mri_noncartesian/GriddingReconGadgetBase.hpp (hunk 6 above)
Comment (xueh2) [reply gadgets/mri_noncartesian/GriddingReconGadgetBase.hpp:98]: Creating array by pointer is obsolete. It leads to the error, as     
[[deprecated]]
explicit hoNDArray(const hoNDArray<T> *a);

File: test/integration/run_gadgetron_test.py (hunk 7 above)
Comment (xueh2) [reply test/integration/run_gadgetron_test.py:622]: the file does not exist at the beginning of test, if running "python run_gadgetron_test.py ...", then os.stat(file) leads to exception.

File: toolboxes/core/cpu/math/hoNDArray_linalg.cpp (hunk 8 above)
Comment (xueh2) [reply toolboxes/core/cpu/math/hoNDArray_linalg.cpp:594]: test added.

File: gadgets/mri_core/generic_recon_gadgets/GenericImageReconGadget.cpp (hunk 5 above)
Comment (xueh2) [reply gadgets/mri_core/generic_recon_gadgets/GenericImageReconGadget.cpp:104]: removed. Vendor features should not be in gadgetron.

**Review by hansenms — APPROVED** [review]

**Pull request merged**